### DISCUSSION TITLE
Lightweight client-like implementation to query a DHT network

### DISCUSSION THREAD

**Discussion by Yashvardhan-A** [savoirfairelinux/opendht#703] [Q&A]
Hey! 

I want to know if it is possible to write a lightweight client in C that would not participate in a DHT network but would only be able to make queries to get stored information. The idea is to be able to deploy this 'client' on constrained devices that could get some kind of data from the network in which more powerful nodes are participating.

Thank you for your time and help.

**Comment by AmarOk1412** [comment]
https://github.com/savoirfairelinux/opendht/wiki/REST-API

Indeed the REST API can be used because you are not doing a node of the network, just querying infos

**Reply by Yashvardhan-A** [reply]
Thank you for your prompt reply!

So as I understand, a (server)node would be required to host the HTTP proxy server that could be used to make REST API calls through a client. However, this would make that particular server a central node to the other lightweight clients. Is it possible to query the network directly by participating as a lightweight node that doesn't store network data (or minimal data) and can still query the network? I would also have to write this client in C for support in embedded devices. I am not familiar with the library code yet and wanted to know if doing something like this is possible before I dive deeper. In case it is possible to do so, what would be a good starting point in the library?

Thank you for your time and help.

**Reply by aberaud** [reply]
There is a "client mode" for that exact purpose, but it's not well documented

**Reply by Yashvardhan-A** [reply]
Thank you for pointing that out. I'll try and use it for my purpose and also contribute to its documentation. Could tell me if it is possible to compile client mode separately as a very minimal, barebones build? If so it could be used for multiple platforms without having library dependency issues.

**Reply by aberaud** [reply]
Just did some digging in the code and figured it was never fully implemented. Adding it now, along with the config flag.

Dependencies like msgpack and gnutls would still be required for the client mode.

**Reply by Yashvardhan-A** [reply]
By fully implemented you mean the CMake side or the code implementation? Well, whichever it is, looking forward to using it. Yeah some dependencies can't be shaken off. I am starting to understand the library based on whatever little experience I have with it till now, could you guide me as how can I go about understanding it better as well where may I start if I wish to contribute to its development.

**Reply by aberaud** [reply]
The code was only implemented on the receiver side, which means that the "client" flag in dht network packets is already respected, but never sent.
I implemented adding the flag on the sender side (allowing an opendht node to act as a "client").
Thanks for the proposal, you can review the pull request here: https://github.com/savoirfairelinux/opendht/pull/706

**Reply by Yashvardhan-A** [reply]
Thanks!
